File: provider/orchestrator/registry.go
package main

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
	"sync"

	"github.com/jolucas/wasm-af/pkg/taskstate"
)

// ParamEnrichment declares a derived parameter computed from an existing
// step param. This moves agent-specific knowledge (e.g. "extract domain
// from URL") out of the orchestrator Go code and into the agent registry JSON.
type ParamEnrichment struct {
	Source    string `json:"source"`    // param name to read from
	Target    string `json:"target"`    // enriched param name to write
	Transform string `json:"transform"` // transform to apply: "domain"
}

// AgentMeta describes a single agent type's metadata: how to load it,
// what capability it requires, what host functions it needs, how to
// build its input payload, and what param enrichments to apply.
type AgentMeta struct {
	WasmName      string            `json:"wasm_name"`
	Capability    string            `json:"capability"`
	ContextKey    string            `json:"context_key"`
	HostFunctions []string          `json:"host_functions"`
	PayloadFields map[string]any    `json:"payload_fields"`
	Enrichments   []ParamEnrichment `json:"enrichments,omitempty"`
	Splice        bool              `json:"splice,omitempty"`
	External      bool              `json:"external,omitempty"`
}

// AgentRegistry holds the set of known agent types. It is safe for
// concurrent use — platform agents are loaded at startup from JSON and
// external (BYOA) agents can be registered/removed at runtime.
type AgentRegistry struct {
	mu     sync.RWMutex
	agents map[string]*AgentMeta
}

// ParseAgentRegistry parses agent registry JSON bytes.
func ParseAgentRegistry(data []byte) (*AgentRegistry, error) {
	var raw map[string]*AgentMeta
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("parse agent registry: %w", err)
	}
	for name, meta := range raw {
		if err := validateAgentMeta(name, meta); err != nil {
			return nil, err
		}
	}
	return &AgentRegistry{agents: raw}, nil
}

func validateAgentMeta(name string, meta *AgentMeta) error {
	if meta.WasmName == "" {
		return fmt.Errorf("agent %q: wasm_name is required", name)
	}
	if meta.Capability == "" {
		return fmt.Errorf("agent %q: capability is required", name)
	}
	if meta.ContextKey == "" {
		return fmt.Errorf("agent %q: context_key is required", name)
	}
	return nil
}

// Get returns the metadata for an agent type, or an error if unknown.
func (r *AgentRegistry) Get(agentType string) (*AgentMeta, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	meta, ok := r.agents[agentType]
	if !ok {
		return nil, fmt.Errorf("unknown agent type %q (not in agent registry)", agentType)
	}
	return meta, nil
}

// Register adds or replaces an agent in the registry. The same
// validation rules that apply at startup are enforced here.
func (r *AgentRegistry) Register(name string, meta *AgentMeta) error {
	if err := validateAgentMeta(name, meta); err != nil {
		return err
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	r.agents[name] = meta
	return nil
}

// Remove deletes an agent from the registry. It is a no-op if the
// agent does not exist.
func (r *AgentRegistry) Remove(name string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.agents, name)
}

// List returns a snapshot of all registered agents keyed by name.
func (r *AgentRegistry) List() map[string]*AgentMeta {
	r.mu.RLock()
	defer r.mu.RUnlock()
	out := make(map[string]*AgentMeta, len(r.agents))
	for k, v := range r.agents {
		out[k] = v
	}
	return out
}

// ListExternal returns a snapshot of only the external (BYOA) agents.
func (r *AgentRegistry) ListExternal() map[string]*AgentMeta {
	r.mu.RLock()
	defer r.mu.RUnlock()
	out := make(map[string]*AgentMeta)
	for k, v := range r.agents {
		if v.External {
			out[k] = v
		}
	}
	return out
}

// ClearExternal removes all agents with External == true.
func (r *AgentRegistry) ClearExternal() {
	r.mu.Lock()
	defer r.mu.Unlock()
	for k, v := range r.agents {
		if v.External {
			delete(r.agents, k)
		}
	}
}

// BuildPayload constructs the JSON payload string for a step, using the
// agent's payload_fields definition. Field values are resolved:
//   - "step.params.<key>" → step.Params[key]
//   - "task.context.<key>" → state.Context[key]
//   - numeric/bool literals → inlined as-is
func BuildPayload(meta *AgentMeta, state *taskstate.TaskState, step *taskstate.Step) string {
	if len(meta.PayloadFields) == 0 {
		return "{}"
	}

	out := make(map[string]any, len(meta.PayloadFields))
	for field, spec := range meta.PayloadFields {
		switch v := spec.(type) {
		case string:
			out[field] = resolveFieldRef(v, state, step)
		default:
			out[field] = v
		}
	}

	b, _ := json.Marshal(out)
	return string(b)
}

func resolveFieldRef(ref string, state *taskstate.TaskState, step *taskstate.Step) any {
	if key, ok := strings.CutPrefix(ref, "step.params."); ok {
		return step.Params[key]
	}
	if key, ok := strings.CutPrefix(ref, "task.context."); ok {
		return state.Context[key]
	}
	if n, err := strconv.ParseFloat(ref, 64); err == nil {
		if n == float64(int(n)) {
			return int(n)
		}
		return n
	}
	return ref
}

File: crates/orchestrator/src/registry.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::RwLock;
use wasm_af_taskstate::{Step, TaskState};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParamEnrichment {
    pub source: String,
    pub target: String,
    pub transform: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentMeta {
    pub wasm_name: String,
    pub capability: String,
    pub context_key: String,
    #[serde(default)]
    pub host_functions: Vec<String>,
    #[serde(default)]
    pub payload_fields: HashMap<String, serde_json::Value>,
    #[serde(default)]
    pub enrichments: Vec<ParamEnrichment>,
    #[serde(default)]
    pub splice: bool,
    #[serde(default)]
    pub external: bool,
}

pub struct AgentRegistry {
    agents: RwLock<HashMap<String, AgentMeta>>,
}

impl AgentRegistry {
    pub fn parse(data: &[u8]) -> Result<Self, anyhow::Error> {
        let raw: HashMap<String, AgentMeta> = serde_json::from_slice(data)?;
        for (name, meta) in &raw {
            validate_agent_meta(name, meta)?;
        }
        Ok(AgentRegistry {
            agents: RwLock::new(raw),
        })
    }

    pub fn get(&self, agent_type: &str) -> Result<AgentMeta, anyhow::Error> {
        let agents = self.agents.read().unwrap();
        agents
            .get(agent_type)
            .cloned()
            .ok_or_else(|| anyhow::anyhow!("unknown agent type {agent_type:?}"))
    }

    pub fn register(&self, name: &str, meta: AgentMeta) -> Result<(), anyhow::Error> {
        validate_agent_meta(name, &meta)?;
        let mut agents = self.agents.write().unwrap();
        agents.insert(name.to_string(), meta);
        Ok(())
    }

    pub fn remove(&self, name: &str) {
        let mut agents = self.agents.write().unwrap();
        agents.remove(name);
    }

    pub fn list(&self) -> HashMap<String, AgentMeta> {
        let agents = self.agents.read().unwrap();
        agents.clone()
    }

    #[allow(dead_code)]
    pub fn list_external(&self) -> HashMap<String, AgentMeta> {
        let agents = self.agents.read().unwrap();
        agents
            .iter()
            .filter(|(_, v)| v.external)
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect()
    }

    #[allow(dead_code)]
    pub fn clear_external(&self) {
        let mut agents = self.agents.write().unwrap();
        agents.retain(|_, v| !v.external);
    }

    pub fn is_platform(&self, name: &str) -> bool {
        let agents = self.agents.read().unwrap();
        agents.get(name).map(|m| !m.external).unwrap_or(false)
    }
}

fn validate_agent_meta(name: &str, meta: &AgentMeta) -> Result<(), anyhow::Error> {
    if meta.wasm_name.is_empty() {
        anyhow::bail!("agent {name:?}: wasm_name is required");
    }
    if meta.capability.is_empty() {
        anyhow::bail!("agent {name:?}: capability is required");
    }
    if meta.context_key.is_empty() {
        anyhow::bail!("agent {name:?}: context_key is required");
    }
    Ok(())
}

/// Construct the JSON payload string for a step using the agent's payload_fields.
pub fn build_payload(meta: &AgentMeta, state: &TaskState, step: &Step) -> String {
    if meta.payload_fields.is_empty() {
        return "{}".to_string();
    }

    let mut out = serde_json::Map::new();
    for (field, spec) in &meta.payload_fields {
        match spec {
            serde_json::Value::String(ref_str) => {
                out.insert(
                    field.clone(),
                    serde_json::Value::String(resolve_field_ref(ref_str, state, step)),
                );
            }
            other => {
                out.insert(field.clone(), other.clone());
            }
        }
    }
    serde_json::to_string(&out).unwrap_or_else(|_| "{}".to_string())
}

fn resolve_field_ref(reference: &str, state: &TaskState, step: &Step) -> String {
    if let Some(key) = reference.strip_prefix("step.params.") {
        return step.params.get(key).cloned().unwrap_or_default();
    }
    if let Some(key) = reference.strip_prefix("task.context.") {
        return state.context.get(key).cloned().unwrap_or_default();
    }
    reference.to_string()
}

/// Apply enrichments (e.g., extract domain from URL) to a copy of step params.
pub fn enrich_params(
    params: &HashMap<String, String>,
    enrichments: &[ParamEnrichment],
) -> HashMap<String, String> {
    let mut out = params.clone();
    for e in enrichments {
        if let Some(src) = out.get(&e.source).cloned() {
            let val = match e.transform.as_str() {
                "domain" => extract_domain(&src),
                _ => src,
            };
            out.insert(e.target.clone(), val);
        }
    }
    out
}

fn extract_domain(raw_url: &str) -> String {
    url::Url::parse(raw_url)
        .ok()
        .and_then(|u| u.host_str().map(|h| h.to_string()))
        .unwrap_or_else(|| raw_url.to_string())
}
